### ISSUE TITLE
Feature request: non-blocking enqueue operations

### ISSUE THREAD

**Issue by leandro-gracia-gil** [tensorflow/tensorflow#12054]
As for version 1.2, queue operations involving enqueuing can block if a queue is full or empty. This is very useful for designing input threads feeding the queue from static datasets as it suspends the thread until it can proceed doing its work. However, this behavior might not be desirable or acceptable if instead the data being fed comes from an asynchronous and continuously changing live source at a best effort basis, because blocking the thread implies missing the continuity of the source.

For example, let's say we have a distributed reinforcement learning environment where an external game plays asynchronously and a thread tries its best to read the most recent state of the game and input the actions to take. As a result, this thread keeps producing training batches that are enqueued in a PaddingFIFOQueue, which are consumed by a separate trainer.

In this case, if for any reason the trainer takes too long or has an un expected peak in the number of training batches it receives, the queue will fill and the producer thread(s) will block. Since the source of data (the external game) keeps playing, this means that the next time the producer threads unblock they will probably have lost temporal coherency on the game. In this case, dropping a training batch would be more desirable than blocking the thread until such batch can be enqueued.

For this reason, I'd like to propose the following feature: a new optional argument to the enqueuing methods (enqueue and enqueue_many) that allows to immediately return when the queue is full instead of blocking the thread. This argument would default to the existing behavior (e.g., 'non_blocking=False'), so no code updates are required.

Signaling if/how many elements were successfully enqueued or failed to do so is likely a desirable output of a non-blocking enqueue operation. However, implementing it might not be trivial without altering the function return signature. If we find this is something we should provide, it might be better to instead of adding a new argument simply add new non-blocking versions of the enqueue ops that return a (op, num_elements_discarded) tuple.

Any comments or suggestions are most welcome.

**Comment by yaroslavvb** [comment]
From implementation standpoint, behavior like this could be implemented by having enqueue kernel return bad status, so that entire session.run call terminates immediately with exception.

However I wonder if this functionality could be implemented at the client level. IE, a client can check how much capacity is left in the queue, and if it's too small, skip the enqueue call altogether. You can include timestamps in your batches and drop batches which are too old. You can use timeout feature to drop any enqueue call that blocks for more than x milliseconds

**Comment by leandro-gracia-gil** [comment]
Thanks for your suggestions!

I've considered them one by one, but as far as I can tell they don't fully solve the problem. Some can offer a partial workaround, but with limitations.

- **Issuing enqueue calls in their own Python thread**: that would be a straightforward solution to the problem, but unfortunately it's not available in this case because the problem enforces us to not block the thread generating the data. We can create yet a new thread to run the enqueue ops, but then we have the very same problem: we need to feed data to the thread doing the enqueue without blocking the thread generating it.

- **Checking how much capacity is left, then skipping enqueuing conditionally**: this should work for as long as we have a single thread feeding the queue. If we happen to have multiple threads, however, we would need some kind of synchronization mechanism to ensure no other threads enqueues between our check and our enqueue op. Is there anything like that already I can use?  (BTW, a minor note: I found myself missing a queue capacity property when giving this a try.)

- **Including timestamp in the batches and dropping the ones that are too old**: I'm not sure if I understood this suggestion correctly. In the enqueuing thread all data is recent, and in the dequeuing side the thread is busy running training steps. The only way I can think to do this would be to have some intermediate dequeuing thread that goes over the queue contents and filters them out. That would require having a secondary queue. Still, that won't help if there's a peak of finished short episodes since the queue would fill but all data would be very recent. I wouldn't ensure blocking cannot happen either.

- **Using the timeout feature to drop enqueue calls that block for more than x milliseconds**: I actually considered doing this before writing the feature request. While this might work, it transforms the problem into finding a machine-dependent time threshold that is as low as possible for not blocking the thread (we're talking about producing data at multiple frames per second, blocking for milliseconds does matter) and high enough to ensure we can still compute the training batches we enqueue when there's enough space in the queue. Since I expect this to be distributed and run across different kinds of machines with possibly different kinds of work loads, having a machine-dependent threshold is something I'd rather avoid as much as possible.

If there's something I'm missing or something I misunderstood please do let me know. Any suggestions or ideas are most welcome.

**Comment by yaroslavvb** [comment]
For bullet 2, there's a practical upper limit on the number of items that could be enqueued by other threads, say k, so the trick would be to have a spare buffer of k+1, and skip the enqueue when there's a risk of freezing. For OpenAI universe project we had to deal with realtime data and managed to deal with stale data using combination of tricks like above

I guess the issue is -- how much work is would add it to implement/maintain such op vs. how much work it would remove to need to do workarounds like above. 

Perhaps one way to proceed would be for the motivated user to implement a user_op that implements a queue that throws exception instead of blocking, since that would give a sense how complicated adding such op is

cc @mrry in case he has any more comments

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by yaroslavvb** [comment]
I'm fine with closing this, seems vague or hard to implement

**Comment by mrry** [comment]
Catching up on old issues: I'm inclined to freeze the queue API as-is, with a view to deprecating it. Most queue-based use cases can be satisfied by the new `tf.data` API, and more advanced use cases (such as this one) would be more naturally implemented using standard Python objects and Eager mode.

**Comment by leandro-gracia-gil** [comment]
While I'm not against closing this, I disagree with the idea that the new tf.data API should deprecate the queue API. While tf.data does a pretty good job to simplify input pipelines, queues can also be used to perform more complex tasks like synchronizing distributed jobs within the TF graph and creating message pipelines across them.

For example, a tf.FIFOQueue can also be used as a distributed semaphore that locks in two ways: when acquiring with value (size) zero and when releasing with value of N (capacity). This can be used to perform mutually exclusive operations in different jobs if needed, creating distributed barriers and so on.

Similarly tf.FIFOQueue and tf.PaddingFIFOQueue can be used to implement a tensor-based messaging system across jobs.

Note that using python synchronization mechanisms is not always an alternative. For example, in my use case all I can use are TF operations restored from a graph def that I call using the C API in Windows. In that case being able to do all the above within the TF graph is a must.

So, please, consider not deprecating queues even if their API is frozen. They do have more uses than just setting up an input pipeline.

**Comment by agrinh** [comment]
Came here to request non-blocking enqueue as well. Currently check the size of the queue as a proxy which doesn't seem ideal.

But have to add yet another request to not deprecate the queue API. We have some applications where it's really useful. Beyond what @leandro-gracia-gil already mentioned some of the simplest implementations of fast-wavenet rely on the `tf.FIFOQueue`. Also seen a use for it in experience-replay like use-cases. However, the QueueRunner etc. I haven't had any use for.
